### PR TITLE
Add tags to container

### PR DESCRIPTION
**Pull request by yiiliveext**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ✔️
| Fixed issues | #125

### CODE REVIEW

File: src/Container.php [hunk 1]
```diff
@@ -115,6 +119,8 @@ protected function delegateLookup(ContainerInterface $container): void
     protected function set(string $id, $definition): void
     {
         $this->validateDefinition($definition);
+        $tags = $this->extractTags($definition);
```
Comment (thenotsoft) [inline src/Container.php:122]:
```suggestion
```

File: src/Container.php [hunk 2]
```diff
@@ -115,6 +119,8 @@ protected function delegateLookup(ContainerInterface $container): void
...
+        $this->setTags($id, $tags);
```
Comment (thenotsoft) [inline src/Container.php:123]:
```suggestion
        $this->setTags($id, $this->extractTags($definition));
```

Comment (yiiliveext) [reply src/Container.php:124]: oudated

File: src/Container.php (hunk 1 above)
Comment (yiiliveext) [reply src/Container.php:121]: outdated

File: src/Container.php [hunk 3]
```diff
@@ -131,18 +138,61 @@ protected function setMultiple(array $config): void
         }
     }
 
+    private function extractDefinition($definition)
+    {
+        if (is_array($definition) && isset($definition['__definition'])) {
```
Comment (samdark) [inline src/Container.php:143]: Do we really need this new syntax?

```php
[
  '__definition' => [
      '__class' => MyClass::class,
      '__construct()' => [42],
  ],
  '__tags' => ['debug', 'users']
]
```

Isn't `__tags` without `__definition` enough?

File: src/Container.php [hunk 4]
```diff
@@ -131,18 +138,61 @@ protected function setMultiple(array $config): void
...
+        }
+
+        return $definition;
+    }
+
+    private function extractTags($definition): array
+    {
+        if (is_array($definition) && isset($definition['__tags']) && is_array($definition['__tags'])) {
+            $tags = $definition['__tags'];
+            unset($definition['__tags']);
```
Comment (samdark) [inline src/Container.php:154]: Why do we need to unset here?

File: src/Container.php [hunk 5]
```diff
@@ -158,6 +208,25 @@ private function build(string $id)
         return $object;
     }
 
+    private function isTagAlias(string $id): bool
+    {
+        return substr($id, 0, 4) === 'tag@';
```
Comment (samdark) [inline src/Container.php:213]:
```suggestion
        return strpos($id, 'tag@') === 0;
```

**Review by samdark — APPROVED** [review]
The idea itself is alright 👍 Need at least some docs to merge it.

File: src/Container.php (hunk 3 above)
Comment (BoShurik) [reply src/Container.php:143]: Maybe use `!tag` to be sure there is no such method

Comment (yiiliveext) [reply src/Container.php:143]: > Do we really need this new syntax?
> 
> ```
> [
>   '__definition' => [
>       '__class' => MyClass::class,
>       '__construct()' => [42],
>   ],
>   '__tags' => ['debug', 'users']
> ]
> ```
> 
> Isn't `__tags` without `__definition` enough?

This is for closure definitions

**Comment by samdark** [comment]
Need docs to get an idea about how it is intended to be used.

File: src/Container.php (hunk 4 above)
Comment (yiiliveext) [reply src/Container.php:154]: Yes, this is for array definitions
```php
[
   '__class' => MyClass::class,
   '__construct()' => [42],
   '__tags' => ['debug', 'users']
]
```

File: src/Container.php (hunk 3 above)
Comment (xepozz) [reply src/Container.php:143]: How about to make the second argument like a stub Definition and in closure we  can set all definitions properties like the following:
```php
Interface::class => function ($container, $definition) {
  $definition->setTags($tags);
  $definition->setAnythingElse($var);
  return new Class();
}
```

**Review by xepozz — COMMENTED** [review]
What's the result will be for calling `$container->has('tag@engine')`?

**Comment by yiiliveext** [comment]
> What's the result will be for calling `$container->has('tag@engine')`?

Good question.

File: src/Container.php (hunk 3 above)
Comment (samdark) [reply src/Container.php:143]: Interesting idea but needs to be considered separately. An issue would be helpful.

Comment (samdark) [reply src/Container.php:143]: > This is for closure definitions

Huh?

File: src/Container.php (hunk 4 above)
Comment (samdark) [reply src/Container.php:154]: But the definition isn't used afterwards and there's no sense in removing anything from it.

Comment (yiiliveext) [reply src/Container.php:154]: It's used afterward, why not?

Comment (samdark) [reply src/Container.php:154]:
```php
private function extractTags($definition): array
{
    if (is_array($definition) && isset($definition['__tags']) && is_array($definition['__tags'])) {
        $tags = $definition['__tags'];
        unset($definition['__tags']);
	// definition isn't returned or used afterwards
        $this->checkTags($tags);
        return $tags;
    }
    return [];
}
```

Comment (yiiliveext) [reply src/Container.php:154]: Oh, yes. Just in the previous version `$definition` was by reference)

**Comment by xepozz** [comment]
Will a composite container collect all definitions from subcontainers?

File: src/Container.php (hunk 5 above)
Comment (yiiliveext) [reply src/Container.php]: Вщту

File: src/Container.php (hunk 4 above)
Comment (yiiliveext) [reply src/Container.php]: Done

File: src/Container.php (hunk 3 above)
Comment (yiiliveext) [reply src/Container.php:143]: > > This is for closure definitions
> 
> Huh?

```php
EngineMarkTwo::class => [
                '__definition' => function () {
                    return new EngineMarkTwo();
                },
                '__tags' => ['engine'],
            ]
```

**Comment by yiiliveext** [comment]
> Will a composite container collect all tags definitions from subcontainers?

Done

**Comment by samdark** [comment]
Ready for review. Docs are needed.

**Comment by hiqsol** [comment]
Please see alternative proposal https://github.com/yiisoft/di/pull/152
It provides almostly the same but is much simpler.

**Comment by samdark** [comment]
@yiiliveext would you please resolve conflicts?

**Comment by yiiliveext** [comment]
> @yiiliveext would you please resolve conflicts?

Done

**Review by samdark — APPROVED** [review]

**Pull request merged**